src/algebra/monoid_algebra.lean

```diff
@@ -632,6 +698,45 @@ f.single_mul_apply_aux r _ _ _ $ λ a, by rw [zero_add]
 
 end misc_theorems
 
+end add_monoid_algebra
+
+/-!
+#### Conversions between `add_monoid_algebra` and `monoid_algebra`
+
+While we were not able to define `add_monoid_algebra k G = monoid_algebra k (multiplicative G)` due
+to definitional inconveniences, we can still show the types are isomorphic.
+-/
+
+/-- The equivalence between `add_monoid_algebra` and `monoid_algebra` in terms of `multiplicative` -/
+protected def add_monoid_algebra.to_multiplicative [semiring k] [add_monoid G] :
+  add_monoid_algebra k G ≃+* monoid_algebra k (multiplicative G) :=
+{ map_mul' := λ x y, by {
+    simp only [finsupp.dom_congr],
+    apply add_monoid_algebra.map_domain_mul',
+    intros p q,
+    -- At this point the pretty-printing of the goal is nonsense, containing `p + q` despite
+    -- the fact there is no obvious `has_add (multiplicative G)`.
+    simp only [←to_add_mul],
+    refl, },
+  ..finsupp.dom_congr multiplicative.to_add }
+
+/-- The equivalence between `monoid_algebra` and `add_monoid_algebra` in terms of `additive` -/
+protected def monoid_algebra.to_additive [semiring k] [monoid G] :
+  monoid_algebra k G ≃+* add_monoid_algebra k (additive G) :=
+{ map_mul' := λ x y, by {
+    simp only [finsupp.dom_congr],
+    apply monoid_algebra.map_domain_mul',
+    intros p q,
+    -- At this point the pretty-printing of the goal is nonsense, containing `p + q` despite
+    -- the fact there is no obvious `has_mul (additive G)`.
+    simp only [of_mul_mul],
+    refl, },
+  ..finsupp.dom_congr additive.of_mul }
```

Comment:
Aah! If you make `multiplicative` irreducible, we see that this line is also abusing type equality.

```
type mismatch at field 'to_fun'
  (finsupp.dom_congr multiplicative.to_add).to_fun
has type
  (multiplicative ?m_1 →₀ ?m_2) → ?m_1 →₀ ?m_2 : Type (max ? ?)
but is expected to have type
  add_monoid_algebra k G → monoid_algebra k (multiplicative G) : Type (max (max u₂ u₁) u₁ u₂)
```

This is the problem -- multiplicative is on the wrong side.

Reply:
```
protected def add_monoid_algebra.of_multiplicative [semiring k] [add_monoid G] :
  monoid_algebra k (multiplicative G) ≃+* add_monoid_algebra k G :=
{ map_mul' := λ x y, by {
    simp only [finsupp.dom_congr],
    sorry
  },
  ..finsupp.dom_congr multiplicative.to_add }
```
This typechecks even if `multiplicative` is irreducible. Note that I switched the order of the equiv! (and hence changed the name)

Reply:
BTW making `multiplicative` irreducible shows that there is also type abuse going on in `lift` (which is already in mathlib) -- for example 
```
type mismatch at application
  ⇑F a
term
  a
has type
  G
but is expected to have type
  multiplicative G
```
error in definition of lift.to_fun. Type abuse like this creates problems down the line (e.g. the earlier problem about p*q=p+q) and I've found that making stuff like `multiplicative` irreducible stops this sort of thing happening.

Reply:
> This is the problem -- multiplicative is on the wrong side.

Thanks for catching this - but this was a mistake in just one of the two definitions.

Reply:
I think this is a case of two wrong make a right. After the `simp only [finsupp.dom_congr],` line, the goal is:
```lean
--                                 VVVVV add_monoid_algebra k G
map_domain ⇑multiplicative.of_add (x * y) =
-- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ monoid_algebra k multiplicative(G)
  map_domain ⇑multiplicative.of_add x * map_domain ⇑multiplicative.of_add y
```
`map_domain` doesn't know about `*_monoid_algebra`, so has signature `(G →₀ k) → (multiplicative G →₀ k)`. At this point, we've done our first sneaky step in type abuse - we're considering the input an `add_monoid_algebra`, and the output a `monoid_algebra` - even though `map_domain_mul'` implies that both input and output are of the same type.

The second wrong is the tricks we already see in `(f)`.

This would all be a lot easier if we could do away with `add_*`, and have `monoid_algebra k G (+)` and `monoid_algebra k G (*)` as our two versions...

Reply:
I'd argue that type abuse is only a problem in definitions. In proofs, it shouldn't matter - as long as the theorem statement plays no tricks, then proof irrelevance means that abuse in the proof doesn't matter, right?

Reply:
Now, I could replace this proof with
```
    simp only [finsupp.dom_congr],
    rw monoid_algebra.mul_def,
    rw add_monoid_algebra.mul_def,
    simp_rw [map_domain_sum, map_domain_single, of_mul_mul],
    rw finsupp.sum_map_domain_index,
    { congr,
      ext a b,
      rw finsupp.sum_map_domain_index,
      { simp },
      { simp [mul_add] } },
    { simp },
    { simp [add_mul] },
```

but that amounts to a copy-paste of `monoid_algebra.map_domain_mul`, with:
* One `monoid_algebra.mul_def` replaced with ` add_monoid_algebra.mul_def`
* `f.map_mul` replaced with `of_mul_mul`

Given the whole point of the `multiplicative` type tag is to avoid duplicate lemmas, is this really worth it?